### PR TITLE
auto-detect intended sector size from DDI disk images, and configure loopback devices for it

### PR DESCRIPTION
**Pull request by poettering**
This adds logic to automatically configure the loopback sector size based on what the DDI we are about to set up is set up for. It will auto-probe the images for a GPT disk label with the various relevant sector sizes, and then use that.

This should make our stuff work much better with 4K sector DDIs.

Kinda related to #26044

### CODE REVIEW

File: src/shared/loop-util.c [hunk 1]
```diff
@@ -220,6 +222,21 @@ static int loop_configure_fallback(int fd, const struct loop_config *c) {
...
+                int z;
+
+                if (ioctl(fd, LOOP_SET_BLOCK_SIZE, (unsigned long) c->block_size) < 0)
+                        log_debug_errno(errno, "Failed to set sector size, ignoring: %m");
+
+                if (ioctl(fd, BLKSSZGET, &z) < 0)
+                        log_debug_errno(errno, "Failed to read sector size: %m");
+
+                if (z < 0 || (unsigned long) z != c->block_size)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Sector size of loopback device doesn't match what we requested, refusing.");
```
Comment (daandemeyer) [inline src/shared/loop-util.c]: `EIO`? `EINVAL` doesn't seem right

File: src/shared/dissect-image.c [hunk 2]
```diff
@@ -107,6 +108,78 @@ int dissect_fstype_ok(const char *fstype) {
...
+        for (uint32_t sz = 512; sz <= 4096; sz <<= 1) {
+                struct gpt_header *p;
+
+                assert(sizeof(sectors) >= sz * 2);
+                p = (struct gpt_header*) (sectors + sz);
+
+                if (memcmp(p->signature, (const char[8]) { 'E', 'F', 'I', ' ', 'P', 'A', 'R', 'T' }, 8) != 0)
+                        continue;
+
+                if (le32toh(p->revision) != UINT32_C(0x00010000))
```
Comment (daandemeyer) [inline src/shared/dissect-image.c]: Does it make sense to hardcode a revision? Could use a comment here probably

File: src/shared/loop-util.c [hunk 3]
```diff
@@ -491,9 +493,36 @@ static int loop_device_make_internal(
...
+        else if (sector_size == UINT32_MAX) {
+                if (S_ISBLK(st.st_mode)) {
+                        int z;
+
+                        /* If the sector size is specified as UINT32_MAX we'll propagate the sector size of
+                         * the underlying block device. */
+                        if (ioctl(fd, BLKSSZGET, &z) < 0)
+                                return -errno;
+                        if (z <= 0) /* Wut? */
+                                return -EIO;
```
Comment (daandemeyer) [inline src/shared/loop-util.c]: Should probably debug log in this case

File: src/shared/loop-util.c [hunk 4]
```diff
@@ -838,6 +869,12 @@ int loop_device_open(
         if (r < 0 && r != -EOPNOTSUPP)
                 return r;
 
+        int sector_size;
+        if (ioctl(fd, BLKSSZGET, &sector_size) < 0)
+                return -errno;
+        if (sector_size <= 0)
+                return -EIO;
```
Comment (daandemeyer) [inline src/shared/loop-util.c]: Ditto

File: src/dissect/dissect.c [hunk 5]
```diff
@@ -601,6 +601,8 @@ static int action_dissect(DissectedImage *m, LoopDevice *d) {
         else if (arg_json_format_flags & JSON_FORMAT_OFF)
                 printf("      Size: %s\n", FORMAT_BYTES(size));
 
+        printf(" Sec. Size: %" PRIu32 "\n", m->sector_size);
```
Comment (daandemeyer) [inline src/dissect/dissect.c:604]: Why abbreviate here? I'd just write sector size in full here

File: src/shared/dissect-image.c [hunk 6]
```diff
@@ -180,6 +180,38 @@ int probe_sector_size(int fd, uint32_t *ret) {
...
+        if (fstat(fd, &st) < 0)
+                return -errno;
+
+        if (S_ISBLK(st.st_mode)) {
+                int z;
+
+                if (ioctl(fd, BLKSSZGET, &z) < 0)
+                        return -errno;
+                if (z <= 0)
+                        return -EIO;
```
Comment (daandemeyer) [inline src/shared/dissect-image.c]: Ditto

**Review by DaanDeMeyer — APPROVED** [review]
LGTM, just some nits

File: src/dissect/dissect.c (hunk 5 above)
Comment (poettering) [reply src/dissect/dissect.c:604]: then the `:` wouldn't be aligned to the `:` of the previous and next lines!

File: src/shared/dissect-image.c (hunk 2 above)
Comment (poettering) [reply src/shared/dissect-image.c]: well, it's the only one defined version, the one we implement. see the wording from the spec:

https://uefi.org/specs/UEFI/2.10/05_GUID_Partition_Table_Format.html#gpt-header

File: src/shared/blockdev-util.c [hunk 7]
```diff
@@ -762,3 +762,15 @@ int blockdev_reread_partition_table(sd_device *dev) {
 
         return 0;
 }
+
+int blockdev_get_sector_size(int fd, uint32_t *ret) {
+        int ssz = 0;
+
```
Comment (bluca) [inline src/shared/blockdev-util.c:768]: assert(ret)

File: src/shared/loop-util.c [hunk 8]
```diff
@@ -220,6 +223,21 @@ static int loop_configure_fallback(int fd, const struct loop_config *c) {
...
 
+        /* If a block size is requested then try to configure it. If that doesn't work, ignore errors, but
+         * afterwards, let's validate what is in effect, and if it doesn't match what we want, fail */
+        if (c->block_size != 0) {
+                uint32_t ssz;
+
+                if (ioctl(fd, LOOP_SET_BLOCK_SIZE, (unsigned long) c->block_size) < 0)
+                        log_debug_errno(errno, "Failed to set sector size, ignoring: %m");
+
+                r = blockdev_get_sector_size(fd, &ssz);
```
Comment (bluca) [inline src/shared/loop-util.c:235]: didn't you intend to skip this if the ioctl fails? or return?

**Comment by poettering** [comment]
made a bunch of changes now, addressing @DaanDeMeyer's points. Since these are a bit more involved and I added another commit unsetting green label. ptal

File: src/shared/loop-util.c (hunk 8 above)
Comment (poettering) [reply src/shared/loop-util.c:235]: nope. if the ioctl fails, then that's noteworthy so that we debug log it. but, we immediately check what the sector size is that is in effect. If this is the right size anyway, then we can just ignore the previous error. Only if it is not the right size anyway we have to fail.

File: src/shared/dissect-image.c [hunk 9]
```diff
@@ -180,6 +180,24 @@ int probe_sector_size(int fd, uint32_t *ret) {
         return 0;   /* indicate we didn't find it */
 }
 
+int probe_sector_size_prefer_ioctl(int fd, uint32_t *ret) {
+        struct stat st;
+
+        assert(fd >= 0);
+        assert(ret);
+
+        /* Just like probe_sector_size_prefer_ioctl(), but if we are looking at a block device, will use the
```
Comment (bluca) [inline src/shared/dissect-image.c]: did you mean `probe_sector_size()`?

File: src/partition/repart.c [hunk 10]
```diff
@@ -1978,19 +1978,33 @@ static int context_load_partition_table(Context *context) {
...
+        if (arg_sector_size > 0)
                 r = fdisk_save_user_sector_size(c, /* phy= */ 0, arg_sector_size);
+        else {
+                uint32_t ssz;
+
+                if (context->backing_fd < 0) {
+                        context->backing_fd = open(context->node, O_RDONLY|O_CLOEXEC);
+                        if (context->backing_fd < 0)
+                                return log_error_errno(errno, "Failed to open device '%s': %m", context->node);
+                }
```
Comment (daandemeyer) [inline src/partition/repart.c:1990]: The whole logic related to opening the backing fd on demand is super complicated to follow. I'm wondering if we can just open it once early enough and remove all these lazy open checks. Let's leave that for a separate PR though.

**Review by bluca — APPROVED** [review]

**Comment by DaanDeMeyer** [comment]
TEST-58-REPART is failing:

```
23:29:51 [   26.675969] testsuite-58.sh[475]: label-id: 1D2CE291-7CCE-4F7D-BC83-FDB49AD74EBD
23:29:51 [   26.676627] testsuite-58.sh[475]: device: /var/tmp/test-repart.CyIa3XLiYj/zzz
23:29:51 [   26.677303] testsuite-58.sh[475]: unit: sectors
23:29:51 [   26.677949] testsuite-58.sh[475]: first-lba: 2048
23:29:51 [   26.678670] testsuite-58.sh[475]: last-lba: 63494
23:29:51 [   26.679316] testsuite-58.sh[475]: sector-size: 512
23:29:51 [   26.679994] testsuite-58.sh[475]: /var/tmp/test-repart.CyIa3XLiYj/zzz1 : start=        2048, size=       20480, type=C12A7328-F81F-11D2-BA4B-00A0C93EC93B, uuid=39107B09-615D-48FB-BA37-C663885FCE67, name="esp"
23:29:51 [   26.680657] testsuite-58.sh[475]: /var/tmp/test-repart.CyIa3XLiYj/zzz2 : start=       22528, size=       20480, type=4F68BCE3-E8CD-4DB1-96E7-FBCAF984B709, uuid=60F33797-1D71-4DCB-AA6F-20564F036CD0, name="root-x86-64", attrs="GUID:59"
23:29:51 [   26.681324] testsuite-58.sh[475]: /var/tmp/test-repart.CyIa3XLiYj/zzz3 : start=       43008, size=       20480, type=8484680C-9521-48C6-9C11-B0720656F69E, uuid=7E3369DD-D653-4513-ADF5-B993A9F20C16, name="usr-x86-64", attrs="GUID:60"'
23:29:51 [   26.682240] testsuite-58.sh[1591]: + set +ex
23:29:51 [   26.682801] testsuite-58.sh[475]: + assert_in '/var/tmp/test-repart.CyIa3XLiYj/zzz3 : start=       43008, size=       20480, type=8484680C-9521-48C6-9C11-B0720656F69E, uuid=7E3369DD-D653-4513-ADF5-B993A9F20C16, name="usr-x86-64", attrs="GUID:60"' 'label: gpt
23:29:51 [   26.683365] testsuite-58.sh[475]: label-id: 1D2CE291-7CCE-4F7D-BC83-FDB49AD74EBD
23:29:51 [   26.684007] testsuite-58.sh[475]: device: /var/tmp/test-repart.CyIa3XLiYj/zzz
23:29:51 [   26.684686] testsuite-58.sh[475]: unit: sectors
23:29:51 [   26.685393] testsuite-58.sh[475]: first-lba: 2048
23:29:51 [   26.686040] testsuite-58.sh[475]: last-lba: 63494
23:29:51 [   26.686704] testsuite-58.sh[475]: sector-size: 512
23:29:51 [   26.687377] testsuite-58.sh[475]: /var/tmp/test-repart.CyIa3XLiYj/zzz1 : start=        2048, size=       20480, type=C12A7328-F81F-11D2-BA4B-00A0C93EC93B, uuid=39107B09-615D-48FB-BA37-C663885FCE67, name="esp"
23:29:51 [   26.688062] testsuite-58.sh[475]: /var/tmp/test-repart.CyIa3XLiYj/zzz2 : start=       22528, size=       20480, type=4F68BCE3-E8CD-4DB1-96E7-FBCAF984B709, uuid=60F33797-1D71-4DCB-AA6F-20564F036CD0, name="root-x86-64", attrs="GUID:59"
23:29:51 [   26.688767] testsuite-58.sh[475]: /var/tmp/test-repart.CyIa3XLiYj/zzz3 : start=       43008, size=       20480, type=8484680C-9521-48C6-9C11-B0720656F69E, uuid=7E3369DD-D653-4513-ADF5-B993A9F20C16, name="usr-x86-64", attrs="GUID:60"'
23:29:51 [   26.689643] testsuite-58.sh[1592]: + set +ex
23:29:51 [   26.690198] testsuite-58.sh[475]: + systemd-detect-virt --quiet --container
23:29:51 [   26.690755] testsuite-58.sh[475]: + cat
23:29:51 [   26.691423] testsuite-58.sh[475]: + cat
23:29:51 [   26.692148] testsuite-58.sh[475]: + cat
23:29:51 [   26.692782] testsuite-58.sh[475]: + systemd-repart --definitions=/tmp/test-repart.kWt0pIA4SA --empty=create --size=auto --seed=750b6cd5c4ae4012a15e7be3c29e6a47 --image=/var/tmp/test-repart.CyIa3XLiYj/zzz /var/tmp/test-repart.CyIa3XLiYj/yyy
23:29:51 [   26.693681] testsuite-58.sh[1598]: Opened '/var/tmp/test-repart.CyIa3XLiYj/zzz' in O_RDONLY access mode, with O_DIRECT enabled.
23:29:51 [   26.694125] testsuite-58.sh[1598]: Failed to set up loopback device for /var/tmp/test-repart.CyIa3XLiYj/zzz: Invalid argument
```

File: src/shared/loop-util.c [hunk 11]
```diff
@@ -492,9 +494,30 @@ static int loop_device_make_internal(
...
+                         * the underlying block device. */
+                        r = blockdev_get_sector_size(fd, &sector_size);
+                else {
+                        assert(S_ISREG(st.st_mode));
+
+                        /* If sector size is specified as UINT32_MAX, we'll try to probe the right sector
+                         * size of the image in question by looking for the GPT partition header at various
+                         * offsets. This of course only works if the image already has a disk label. */
+                        r = probe_sector_size(fd, &sector_size);
+                }
```
Comment (daandemeyer) [inline src/shared/loop-util.c]: If you move the prefer_ioctl() commit earlier you can use that here I think

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: src/shared/loop-util.c (hunk 11 above)
Comment (poettering) [reply src/shared/loop-util.c]: but that would mean we'd have to add another fstat()!

Comment (daandemeyer) [reply src/shared/loop-util.c]: You could add a third argument that takes a struct stat pointer so that it only stats if needed I guess? I dunno, don't have a big opinion on this

Comment (poettering) [reply src/shared/loop-util.c]: also, as it turns out we have to operate on a separate fd here with O_DIRECT turned off. The CI failure was caused by trying to run probe_sector_size() with O_DIRECT on. hence this code will change quite a bit once that is addressed.

**Comment by poettering** [comment]
Force pushed a new version. Only change is to the [loop-util: always tell kernel explicitly about loopback sector size](https://github.com/systemd/systemd/pull/26092/commits/22ee78a8987f29e7f837efab86ed090ab78c1170) commit, regarding O_DIRECT handling.

**Review by DaanDeMeyer — APPROVED** [review]
